include/rocksdb/sst_file_writer.h

```diff
@@ -110,53 +110,64 @@ class SstFileWriter {
+  //           key according to the comparator.
   // REQUIRES: comparator is *not* timestamp-aware.
   Status Put(const Slice& user_key, const Slice& value);
 
   // Add a Put (key with timestamp, value) to the currently opened file
-  // REQUIRES: key is after any previously added key according to the
-  // comparator.
-  // REQUIRES: the timestamp's size is equal to what is expected by
-  // the comparator.
+  // REQUIRES: user_key is after any previously added point (Put/Merge/Delete)
+  //           key according to the comparator.
+  // REQUIRES: timestamp's size is equal to what is expected by the comparator.
   Status Put(const Slice& user_key, const Slice& timestamp, const Slice& value);
 
   // Add a Merge key with value to currently opened file
-  // REQUIRES: key is after any previously added key according to comparator.
+  // REQUIRES: user_key is after any previously added point (Put/Merge/Delete)
+  //           key according to the comparator.
   // REQUIRES: comparator is *not* timestamp-aware.
   Status Merge(const Slice& user_key, const Slice& value);
 
   // Add a deletion key to currently opened file
-  // REQUIRES: key is after any previously added key according to comparator.
+  // REQUIRES: user_key is after any previously added point (Put/Merge/Delete)
+  //           key according to the comparator.
   // REQUIRES: comparator is *not* timestamp-aware.
   Status Delete(const Slice& user_key);
 
   // Add a deletion key with timestamp to the currently opened file
-  // REQUIRES: key is after any previously added key according to the
-  // comparator.
-  // REQUIRES: the timestamp's size is equal to what is expected by
-  // the comparator.
+  // REQUIRES: user_key is after any previously added point (Put/Merge/Delete)
+  //           key according to the comparator.
+  // REQUIRES: timestamp's size is equal to what is expected by the comparator.
   Status Delete(const Slice& user_key, const Slice& timestamp);
 
   // Add a range deletion tombstone to currently opened file. Such a range
-  // deletion tombstone does NOT delete other (point) keys in the same file.
+  // deletion tombstone does NOT delete point (Put/Merge/Delete) keys in the
+  // same file.
+  //
+  // Range deletion tombstones may be added in any order, both with respect to
+  // each other and with respect to the point (Put/Merge/Delete) keys in the
+  // same file.
   //
   // REQUIRES: The comparator orders `begin_key` at or before `end_key`
   // REQUIRES: comparator is *not* timestamp-aware.
   Status DeleteRange(const Slice& begin_key, const Slice& end_key);
 
   // Add a range deletion tombstone to currently opened file. Such a range
-  // deletion tombstone does NOT delete other (point) keys in the same file.
+  // deletion tombstone does NOT delete point (Put/Merge/Delete) keys in the
+  // same file.
+  //
+  // Range deletion tombstones may be added in any order, both with respect to
```

Comment:
Do we want to also update comment for BlockBuilder::Add? https://github.com/facebook/rocksdb/blob/bd80433c73691031ba7baa65c16c63a83aef201a/table/block_based/block_builder.h#L40 `SstFileWriter::DeleteRange()` calls it to add out-of-order range tombstones but the contract says keys should be ordered.

Reply:
Huh, I never noticed that. We used to record range tombstones in snapshot stripe order, which is why our read logic needs to be able to handle range tombstones out of (key) order. I suspect we were unknowingly using BlockBuilder in an off-label way then. I don't see any downside to removing that restriction, do you?

Reply:
Done, though LMK if you disagree with removing it entirely

Reply:
It's kind of confusing with Block (which does not appear to require ordering), BlockIter (which does appear to require ordering), etc. I brought back the restriction but mentioned it doesn't apply to range tombstones. Maybe we can think of something better.

Reply:
Where do we require ordering for BlockIter? I think Block and BlockIter can work out-of-order (as in the case of range tombstones) as long as producer and consumer sides agree on the ordering. So maybe the restriction is just a widely used assumption that is broken only by range tombstones?

Reply:
Here the BlockIter says the Block should be sorted: https://github.com/facebook/rocksdb/blob/43e9a60bb24ac8079e182177a6721d56e4fdd11c/table/block_based/block.h#L254

Of course we still use it unsorted (through DataBlockIter), but we have to be careful not to call Seek()...

Reply:
I see. Yeah, good point, only a subset of Block/BlockIter APIs can be used when the block unsorted.